quil-rs/src/program/graph.rs

```diff
@@ -691,3 +691,154 @@ impl<'a> ScheduledProgram<'a> {
+                    Instruction::Pragma(Pragma {
+                        name, arguments, ..
+                    }) if name == RAW_INSTRUCTION => Some(Some({
+                        let frame_condition = if arguments.is_empty() {
+                            FrameMatchCondition::All
+                        } else {
+                            FrameMatchCondition::AnyOfNames(
+                                arguments
+                                    .iter()
+                                    .filter_map(|arg| match arg {
+                                        PragmaArgument::Identifier(name) => Some(name.as_str()),
+                                        PragmaArgument::Integer(_) => None,
+                                    })
+                                    .collect(),
+                            )
+                        };
+
+                        eprintln!("{frame_condition:?}");
+                        let used = program
+                            .frames
+                            .get_matching_keys_for_condition(frame_condition);
+
+                        MatchedFrames {
+                            used,
+                            blocked: HashSet::new(),
+                        }
+                    })),
+                    _ => None,
+                })
+                .set_memory_accesses(|instruction| match instruction {
+                    Instruction::Pragma(Pragma { name, .. }) if name == NO_OP => {
+                        Some(MemoryAccesses::default())
+                    }
+                    Instruction::Pragma(Pragma { name, .. }) if name == RAW_INSTRUCTION => Some({
+                        MemoryAccesses {
+                            captures: HashSet::new(),
+                            reads: [String::from("ro")].into(),
+                            writes: HashSet::new(),
+                        }
+                    }),
+                    _ => None,
+                })
+        }
+
+        #[test]
+        fn test_instructionblock_build() {
+            let program = r#"
+DEFFRAME 0 "quux":
+    SAMPLE-RATE: 1.0
+    INITIAL-FREQUENCY: 1e8
+PRAGMA NO-OP
+PRAGMA RAW-INSTRUCTION
+PRAGMA RAW-INSTRUCTION
+PRAGMA NO-OP
+PRAGMA RAW-INSTRUCTION
+"#
+            .parse::<Program>()
+            .unwrap();
+
+            let instructions: Vec<_> = program.instructions.iter().collect();
+            let terminator = None;
+            let mut custom_handler = get_custom_handler();
+
+            let block = InstructionBlock::build(
+                instructions.clone(),
+                terminator,
+                &program,
+                &mut custom_handler,
+            )
+            .unwrap();
+
+            assert_eq!(block.instructions, instructions);
+            assert_eq!(block.terminator, BlockTerminator::Continue);
+            assert_debug_snapshot!(block.graph);
+        }
+
+        #[test]
+        fn test_instructionblock_build_multiple_gates() {
```

Comment:
Would also be wise to test interleaving with normal `PULSE` instructions, because it's conceivable that a bug would properly schedule pragmas against one another but not properly schedule against "normally"-scheduled instructions